src/couch/src/couch_db_split.erl

```
@@ -0,0 +1,491 @@
+    #state{source_db = SourceDb, targets = Targets} = State,
+    {Key, Target} = pick_target(Id, State, Targets),
+    FDI1 = process_fdi(FDI, SourceDb, Target#target.db),
+    MaxBuffer = State#state.max_buffer_size,
+    Target1 = acc_and_flush(FDI1, Target, MaxBuffer, fun commit_docs/1),
+    {ok, State#state{targets = Targets#{Key => Target1}}}.
+
+
+commit_docs(#target{buffer = [], db = Db} = Target) ->
+    Target#target{db = Db};
+
+commit_docs(#target{buffer = FDIs, db = Db} = Target) ->
+    Pairs = [{not_found, FDI} || FDI <- lists:reverse(FDIs)],
+    {ok, Db1} = couch_db_engine:write_doc_infos(Db, Pairs, []),
+    {ok, Db2} = couch_db_engine:commit_data(Db1),
+    Target#target{buffer = [], buffer_size = 0, db = Db2}.
+
+
+process_fdi(FDI, SourceDb, TargetDb) ->
+    #full_doc_info{id = Id, rev_tree = RTree} = FDI,
+    Acc = #racc{id = Id, source_db = SourceDb, target_db = TargetDb},
+    {NewRTree, NewAcc} = couch_key_tree:mapfold(fun revtree_cb/4, Acc, RTree),
+    {Active, External} = total_sizes(NewAcc),
+    FDI#full_doc_info{
+        rev_tree = NewRTree,
+        sizes = #size_info{active = Active, external = External}
+    }.
+
+
+revtree_cb(_Rev, _Leaf, branch, Acc) ->
+    {[], Acc};
+
+revtree_cb({Pos, RevId}, Leaf, leaf, Acc) ->
+    #racc{id = Id, source_db = SourceDb, target_db = TargetDb} = Acc,
+    #leaf{deleted = Deleted, ptr = Ptr} = Leaf,
+    Doc0 = #doc{
+        id = Id,
+        revs = {Pos, [RevId]},
+        deleted = Deleted,
+        body = Ptr
+    },
+    Doc1 = couch_db_engine:read_doc_body(SourceDb, Doc0),
+    #doc{body = Body, atts = AttInfos0} = Doc1,
+    External = if not is_binary(Body) -> ?term_size(body); true ->
+        couch_compress:uncompressed_size(Body)
+    end,
+    AttInfos = if not is_binary(AttInfos0) -> AttInfos0; true ->
+        couch_compress:decompress(AttInfos0)
+    end,
+    Atts = [process_attachment(Att, SourceDb, TargetDb) || Att <- AttInfos],
```

We checked that process_attachment will raise an exception right? Its only between the `||` and `<-` that badmatch exceptions are swallowed?

---

Yes. If process_attachment blows the whole thing blow up.

```
1> [throw(bam) || true].
** exception throw: bam
2>
```

---

Right, thanks.